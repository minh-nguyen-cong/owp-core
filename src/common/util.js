import debounce from 'lodash/debounce';
import every from 'lodash/every';
import get from 'lodash/get';
import includes from 'lodash/includes';
import isEmpty from 'lodash/isEmpty';
import throttle from 'lodash/throttle';
import moment from 'moment';
import { query } from 'owp/api';
import { setCommonCodeData } from 'owp/auth/store/actions';
import store from 'owp/store';
import * as Actions from 'owp/store/actions';

const REST_API_SUCCESS_MESSAGE = ['처리되었습니다.', '수정되었습니다.', '등록되었습니다.'];
const REST_API_SUCCESS_RESULT_CODE = ['STATUS_1'];

const DATE_LENGTH_COUNTS = [
    'YYYY'.length,
    'YYYY-MM'.length,
    'YYYYMM'.length,
    'YYYY-MM-DD'.length,
    'YYYYMMDD'.length,
    'YYYY-MM-DD HH:mm'.length,
    'YYYY-MM-DD HH:mm:ss'.length,
];

const _dispatch = throttle((action) => {
    store.dispatch(action);
}, 1000);

export function showMessage(messageData = {}) {
    store.dispatch(Actions.hideMessage());

    debounce(() => {
        _dispatch(
            Actions.showMessage({
                anchorOrigin: {
                    vertical: 'top',
                    horizontal: 'center',
                },
                ...messageData,
            })
        );
    }, 100)();
}

export const setCommonCodes = async (commonCodes) => {
    try {
        if (!isEmpty(commonCodes)) {
            _dispatch(setCommonCodeData(commonCodes));
            return;
        }

        const _commonCodes = await query({
            url: `${process.env.REACT_APP_REST_API_URL}/listIPX_CommonCodeTotalAC`,
        });

        _dispatch(setCommonCodeData(_commonCodes));
    } catch (error) {
        console.error(error);
    }
};

export const isJSON = (str) => {
    try {
        JSON.parse(str);
    } catch (e) {
        return false;
    }
    return true;
};

export const sortByTarget = (target = [], path = '') => (a, b) =>
    target.indexOf(get(a, path, a)) - target.indexOf(get(b, path, b));

export const convertNumString = (num) => parseInt(num).toString();

export const makeNowYYYY = () => moment().format('YYYY');
export const makeNowYYYYMMDD = (sep = '-') => moment().format(`YYYY${sep}MM${sep}DD`);
export const makeNowYYYYMMDDHHmm = (dateSep = '-', timeSep = ':') =>
    moment().format(`YYYY${dateSep}MM${dateSep}DDTHH${timeSep}mm`);

export const makeNowYYYYMM = (sep = '-') => moment().format(`YYYY${sep}MM`);

export const makeNowHHmm = (sep = ':') => moment().format(`HH${sep}mm`);

export const makeDateObject = (date = '') => {
    if (date instanceof Date) {
        return date;
    }

    if (includes(DATE_LENGTH_COUNTS, date.length)) {
        return new Date(date);
    }

    return new Date();
};

export const makeDateString = (date = '', { separator = '', useKr = true } = {}) => {
    if (!date || !includes(DATE_LENGTH_COUNTS, date.length)) {
        return date;
    }

    const [yyyy, mm, dd] =
        date.indexOf('-') === -1
            ? [date.substring(0, 4), date.substring(4, 6), date.substring(6, 8)]
            : date.split('-');

    return `${yyyy}${useKr ? '년 ' : separator}${
        useKr ? `${convertNumString(mm)}월` : `${mm}${!isEmpty(dd) ? separator : ''}`
    }${!isEmpty(dd) ? (useKr ? ` ${convertNumString(dd)}일` : dd) : ''}`;
};

export const validateRestApiResponse = (res) =>
    every(
        [
            { key: 'resultCode', values: REST_API_SUCCESS_RESULT_CODE },
            { key: 'resultMessage', values: REST_API_SUCCESS_MESSAGE },
        ].map(({ key, values }) => includes(values, get(res, key)))
    );

export const deepKeysByObject = (obj = {}) =>
    Object.keys(obj)
        .filter((key) => obj[key] instanceof Object)
        .map((key) => deepKeysByObject(obj[key]).map((k) => `${key}.${k}`))
        .reduce((x, y) => x.concat(y), Object.keys(obj))
        .filter((key) => !(get(obj, key) instanceof Object));
