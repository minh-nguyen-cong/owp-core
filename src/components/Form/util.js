export function mapDataToOwpPresetFormProps(options = {}) {
    return (data, { query = {} } = {}) => {
        const dataKey = query.dataKey || options.dataKey || 'items';
        const labelKey = query.labelKey || options.labelKey || 'label';
        const valueKey = query.valueKey || options.valueKey || 'value';

        const hasItemKeysOfOptions = typeof options.item === 'object';
        const hasItemKeysOfQuery = typeof query.item === 'object';

        return {
            [dataKey]: data.map((item = {}) => ({
                [labelKey]:
                    (hasItemKeysOfOptions && item[options.item.labelKey]) ||
                    (hasItemKeysOfQuery && item[query.item.labelKey]) ||
                    item.CODENM ||
                    item['IPX_COMMONCODE.CODENM'],
                [valueKey]:
                    (hasItemKeysOfOptions && item[options.item.valueKey]) ||
                    (hasItemKeysOfQuery && item[query.item.valueKey]) ||
                    item.CODEID ||
                    item['IPX_COMMONCODE.CODEID'],
            })),
        };
    };
}
