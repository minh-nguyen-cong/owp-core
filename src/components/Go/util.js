function makeGoTimelineNodeColor(sid) {
    // 주의
    if (sid === 'D012002') {
        return { event: { stroke: '#000000' }, shape: { fill: '#FFFF00' } };
    }

    // 심각
    if (sid === 'D012003') {
        return { shape: { fill: '#FF0000' } };
    }

    return { shape: { fill: '#0000FF' } };
}

export const makeGoTimelineModel = (data = []) =>
    data.map((node = {}) => ({
        event: node['OWP_ProcessAlarm.AlarmNM'],
        color: makeGoTimelineNodeColor(node['OWP_ProcessAlarm.AlarmSID']),
        date: node['OWP_ProcessAlarm.ADATE'],
    }));
