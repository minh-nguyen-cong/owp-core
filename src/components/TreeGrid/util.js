// TODO: convert it to ramda style function
import isPlainObject from 'lodash/isPlainObject';
import mapKeys from 'lodash/mapKeys';
import {
    assoc,
    complement,
    cond,
    curry,
    filter,
    has,
    isNil,
    keys,
    map,
    merge,
    mergeAll,
    omit,
    path,
    pathEq,
    pipe,
    prop,
    props,
    reduce,
    toPairs,
} from 'ramda';

const isNotNil = complement(isNil);
const renameKeys = curry((keysMap, obj) =>
    reduce(
        (acc, key) => assoc(keysMap[key] || key, obj[key], acc),
        {},
        keys(obj)
    )
);
const makeJsonFromXmlNodes = nodes => {
    try {
        const newItem = {};
        for (const attr of nodes.attributes) {
            newItem[attr.name] = attr.value;
        }

        return newItem;
    } catch (error) {
        return undefined;
    }
};

export const transform = (
    obj = {},
    data = [],
    {
        columnOptions = {},
        cellOptions = {},
        headOptions = [],
        selectColumns = {},
        radioColumns = {},
        suggestColumns = {},
        layoutColumns,
        selectCells,
        radioCells,
    } = {}
) => {
    const header = { ...obj };

    const cols = map(
        column =>
            mergeAll([
                column,
                prop(column.Name, columnOptions),
                prop(column.Name, suggestColumns),
                makeEnumCell(selectColumns[column.Name], 'Enum'),
                makeEnumCell(radioColumns[column.Name], 'Radio'),
            ]),
        Array.isArray(layoutColumns)
            ? layoutColumns
            : pipe(
                  keys,
                  map(key => ({ Name: key, RelWidth: 1 }))
              )(obj)
    );

    const bodyItem = reduce((newItem, [key, value]) => {
        // const newKey = ifElse(
        //     arr => gt(length(arr), 1),
        //     arr => join('.')(tail(arr)),
        //     head
        // )(key.split('.'));
        const newKey = key;

        if (isPlainObject(value)) {
            newItem = value;
        } else {
            newItem[newKey] = value;
        }

        if (cellOptions[newKey]) {
            newItem = mapToValidKeys(newKey, newItem, cellOptions[newKey]);
        }

        if (selectCells && selectCells[newKey]) {
            const selectEnumCell = makeEnumCell(selectCells[newKey], 'Enum');
            newItem = mapToValidKeys(newKey, newItem, selectEnumCell);
        }

        if (radioCells && radioCells[newKey]) {
            const radioEnumCell = makeEnumCell(radioCells[newKey], 'Radio');
            newItem = mapToValidKeys(newKey, newItem, radioEnumCell);
        }

        return newItem;
    });

    const body = map(item => bodyItem({}, toPairs(item)), data);

    return {
        Cols: cols,
        Header: header,
        Head: headOptions,
        Body: [body],
    };
};

const makeEnumCell = (cell, enumType = 'Enum') => {
    if (!cell) return;

    const enumCell = transformToEnum([cell.labelKey, cell.valueKey], cell.data);
    enumCell.Type = enumType;

    return enumCell;
};

const mapToValidKeys = (baseKey, base, item) =>
    merge(
        base,
        mapKeys(item, (value, key) => {
            return `${baseKey}${key}`;
        })
    );

export const transformToEnum = (keys, data) => {
    const pairs = map(props(keys), data);
    const toValidEnumFormat = (o, cur) => {
        o.Enum = `${o.Enum}|${cur[0]}`;
        o.EnumKeys = `${o.EnumKeys}|${cur[1]}`;
        return o;
    };

    return reduce(toValidEnumFormat, { Enum: '', EnumKeys: '' }, pairs);
};

export const parseXML = xmlString => {
    if (typeof DOMParser !== 'function') {
        return;
    }

    return new DOMParser().parseFromString(xmlString, 'text/xml');
};

export const changesAsJSON = pipe(
    parseXML,
    path(['firstChild', 'childNodes']),
    map(makeJsonFromXmlNodes)
);

export const getSelectedData = pipe(
    (data, { rowIdKey } = {}) =>
        map(
            cond([
                [
                    has('Selected'),
                    pipe(
                        omit(['Selected']),
                        renameKeys({ id: rowIdKey })
                    ),
                ],
            ]),
            data
        ),
    filter(isNotNil)
);

export const makeBatchData = pipe(
    (data, { rowIdKey, userSeq } = {}) =>
        map(
            cond([
                [
                    has('Deleted'),
                    pipe(
                        omit(['Deleted']),
                        renameKeys({ id: rowIdKey }),
                        merge({ STATUS: 'Delete' })
                    ),
                ],
                [
                    has('Changed'),
                    pipe(
                        omit(['Changed']),
                        renameKeys({ id: rowIdKey }),
                        merge({ STATUS: 'Update' })
                    ),
                ],
                [
                    has('Added'),
                    pipe(
                        omit([
                            'Added',
                            'id',
                            'Def',
                            'Next',
                            'Parent',
                            'Prev',
                            // pkName, // TODO: https://github.com/leeinbae/owp/issues/42#issuecomment-463965528
                        ]),
                        merge({
                            STATUS: 'Create',
                            [rowIdKey
                                ? `${rowIdKey.split('.')[0]}.WUSERSEQ`
                                : 'WUSERSEQ']: userSeq,
                        })
                    ),
                ],
            ]),
            data
        ),
    filter(isNotNil)
);

export const makeColumnsFromGridLayout = pipe(
    parseXML,
    path(['firstChild', 'childNodes']),
    filter(pathEq(['nodeName'], 'Cols')),
    path(['0', 'childNodes']),
    map(makeJsonFromXmlNodes),
    filter(isNotNil)
);

export const fetchAsText = (url, options = {}) =>
    new Promise((resolve, reject) => {
        fetch(url, options)
            .then(response => resolve(response.text()))
            .catch(error => reject(error));
    });
