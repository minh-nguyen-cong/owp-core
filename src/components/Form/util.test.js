import { mapDataToOwpPresetFormProps } from './util';

describe('mapDataToOwpPresetFormProps', () => {
    const mockData = [
        {
            'IPX_CommonCode.GROUPNM': '판넬_규격',
            'IPX_CommonCode.GROUPID': 'D004000',
            'IPX_CommonCode.FLAG': 'Y',
            'IPX_CommonCode.CODEID': 'D004001',
            'IPX_CommonCode.DESCRIPTION': '',
            'IPX_CommonCode.CODENM': '0.45T',
        },
        {
            'IPX_CommonCode.GROUPNM': '판넬_규격',
            'IPX_CommonCode.GROUPID': 'D004000',
            'IPX_CommonCode.FLAG': 'Y',
            'IPX_CommonCode.CODEID': 'D004002',
            'IPX_CommonCode.DESCRIPTION': '',
            'IPX_CommonCode.CODENM': '0.4T',
        },
    ];

    test('if any options or props are not provided, it should return default array of object', () => {
        const items = mapDataToOwpPresetFormProps()(mockData);
        expect(items).toMatchSnapshot();
    });

    test('if options is provided, it should return valid array of object based on `options`', () => {
        const options = {
            dataKey: 'suggestions',
            item: {
                labelKey: 'IPX_CommonCode.GROUPNM',
                valueKey: 'IPX_CommonCode.GROUPID',
            },
        };
        const items = mapDataToOwpPresetFormProps(options)(mockData);
        expect(items).toMatchSnapshot();
    });

    test('if options and props are provided, it should return valid array of object based on `props`', () => {
        const options = {
            dataKey: 'suggestions',
            item: {
                labelKey: 'IPX_CommonCode.GROUPNM',
                valueKey: 'IPX_CommonCode.GROUPID',
            },
        };
        const props = {
            query: {
                dataKey: 'items',
                item: {
                    labelKey: 'IPX_CommonCode.CODENM',
                    valueKey: 'IPX_CommonCode.CODEID',
                },
            },
        };
        const items = mapDataToOwpPresetFormProps(options)(mockData, props);
        expect(items).toMatchSnapshot();
    });

    test('if options and props are provided, it should return valid array of object based on `props` (with root label, value key)', () => {
        const options = {
            dataKey: 'suggestions',
            labelKey: 'Text',
            valueKey: 'Name',
            item: {
                labelKey: 'IPX_CommonCode.GROUPNM',
                valueKey: 'IPX_CommonCode.GROUPID',
            },
        };
        const props = {
            query: {
                dataKey: 'items',
                labelKey: 'Text',
                valueKey: 'Name',
                item: {
                    labelKey: 'IPX_CommonCode.CODENM',
                    valueKey: 'IPX_CommonCode.CODEID',
                },
            },
        };
        const items = mapDataToOwpPresetFormProps(options)(mockData, props);
        expect(items).toMatchSnapshot();
    });
});
