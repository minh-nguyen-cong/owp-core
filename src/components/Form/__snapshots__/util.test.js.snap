// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDataToOwpPresetFormProps if any options or props are not provided, it should return default array of object 1`] = `
Object {
  "items": Array [
    Object {
      "label": "0.45T",
      "value": "D004001",
    },
    Object {
      "label": "0.4T",
      "value": "D004002",
    },
  ],
}
`;

exports[`mapDataToOwpPresetFormProps if options and props are provided, it should return valid array of object based on \`props\` (with root label, value key) 1`] = `
Object {
  "items": Array [
    Object {
      "Name": "D004000",
      "Text": "판넬_규격",
    },
    Object {
      "Name": "D004000",
      "Text": "판넬_규격",
    },
  ],
}
`;

exports[`mapDataToOwpPresetFormProps if options and props are provided, it should return valid array of object based on \`props\` 1`] = `
Object {
  "items": Array [
    Object {
      "label": "판넬_규격",
      "value": "D004000",
    },
    Object {
      "label": "판넬_규격",
      "value": "D004000",
    },
  ],
}
`;

exports[`mapDataToOwpPresetFormProps if options is provided, it should return valid array of object based on \`options\` 1`] = `
Object {
  "suggestions": Array [
    Object {
      "label": "판넬_규격",
      "value": "D004000",
    },
    Object {
      "label": "판넬_규격",
      "value": "D004000",
    },
  ],
}
`;
